### ISSUE TITLE
Drop support for FreeBSD 10 from std

### ISSUE THREAD

**Issue by asomers** [rust-lang/rust#89058]
Rust's standard library currently uses a FreeBSD 10 compatible ABI.  But FreeBSD 10 went EoL on October 3, 2017.  And now FreeBSD 11 will itself be EoL on September 30, 2021.  Isn't it finally time to drop FreeBSD 10 support and update the standard library's ABI to FreeBSD 12, or at least 11?

**Comment by nagisa** [comment]
Ultimately users of the platform are best positioned to answer this question. There are probably quite few casual users of FreeBSD around here, so somebody who is actively involved in the FreeBSD community would likely be best positioned to provide some guidance here.

**Comment by asomers** [comment]
Well, I'm a FreeBSD developer.  And I can tell you that we're very proud of our backwards-compatibility, but it doesn't require maintaining new tools for old OS versions.  Instead, if a person must use FreeBSD 10, the best way to build new software for it would be to use a snapshot or fork of an old version of the Ports tree.  That would include an older version of rustc, one that was known to work with FreeBSD 10.  In fact, the ports tree itself quickly drops support for old OS versions (it's in the process of dropping support for 11 right now).

**Comment by cemeyer** [comment]
I’m also a FreeBSD and Rust developer and would support dropping 10 support immediately, and 11 support in a few days, when it is officially EOL.

**Comment by delphij** [comment]
I'm also a FreeBSD developer and I would second @cemeyer . FreeBSD 10 support should be gone as it's EoL'ed almost 3 years ago, and FreeBSD 11 support would be ended [really soon](https://www.freebsd.org/security/) too.

At FreeBSD we are committed to provide backward compatibility for applications compiled for older version of FreeBSD releases by maintaining stable system call interfaces; for C runtime we are providing versioned symbols, so legacy binaries would call the old interfaces, which are usually wrappers of new interfaces.  While we are proud that we have provided compatibility shims which worked well for the old binaries, we generally encourage application developers to adopt newer API to take full advantage of the performance improvements and flexibility on newer FreeBSD releases.

**Comment by laffer1** [comment]
Dropping FreeBSD 10 support is likely fine at this point.  I think it's premature with 11, despite the upcoming EOL date.

**Comment by bsdimp** [comment]
Dropping FreeBSD 10 is fine. There have been no security updates to it in over 3 years, so nobody with any kind of security concerns are still running it. FreeBSD 11 is about to be in the same boat. Unless there's some really compelling application stuck on FreeBSD 11 that can't recompile and redeploy to FreeBSD 12, it would be far more beneficial to update to FreeBSD 12's ABI. Old ABIs are generally only supported on a best-effort basis. Given how far out of date FreeBSD 10 at this point, that best effort decayed to zero long ago. 

I believe that community survey has shown that people have completely moved on from FreeBSD 10.

**Comment by bapt** [comment]
Following the freebsd EOL policy is perfectly fine. FreeBSD 10 support should be dropped now, FreeBSD 11 should be dropped by November 1rst.

**Comment by unrelentingtech** [comment]
12 is about the only real backwards-incompatible break (64-bit inodes) in a long long time. EoL of everything before that is a great event \o/ Rust definitely should start the move into the ino64 world soon (by the end of the year). IIRC the libc crate already has flags to use newer ABI, so they should become the default.

**Comment by dch** [comment]
as asomers points out, the FreeBSD ports tree (from which all the packages for end users are built), drops support at the same time. Users who don't stay current with a supported FreeBSD release, will stay on the last rust version that they installed, without a path for upgrade.

**Comment by MikaelUrankar** [comment]
Drop support for FreeBSD 10 and 11  please.

**Comment by asomers** [comment]
@mirabilos could you please explain your objection?

**Comment by emaste** [comment]
There's a related discussion happening for Go now, although Go's avoidance of libc makes it interesting: https://github.com/golang/go/issues/48164

I agree with dropping FreeBSD 10 support immediately and also agree with moving to 12.x the minimum supported version. However, the official EOL is hust over a week from the time I'm writing this comment and I would suggest waiting a short time after that, to allow time for a heads-up on FreeBSD mailing lists or elsewhere.

There are a number of companies and others building products and projects on top of FreeBSD who for better or worse continue to use older, unsupported versions. While it is not reasonable (or possible) to accommodate this case long-term, I'd generally like to avoid forcing them into an urgent and unplanned need to upgrade.

**Comment by allanjude** [comment]
Specifically, getting to to at least the 12 ABI gets rust onto the good side of the changes to support 64-bit inodes and longer filesystem mountpoints.

**Comment by bdrewery** [comment]
TLDR, Yes, but just removing everything will cause problems on 12+. Builders need to be updated to 12 first. A lot of these symbols are from 7.0...

The issue here is far more complex than it seems and discussed more at https://github.com/rust-lang/libc/issues/570.
This post is long because the premise here is (unintentionally) misleading to those who are not aware of the details. There are symbols bound to versions from [FreeBSD 7.0](https://github.com/rust-lang/libc/pull/937/files#diff-a828924ffbc30c564e49bc3652ffa5375a123671b228fc99b5775a29cff6eea5R421) when [symbol versioning was introduced](https://github.com/freebsd/freebsd-src/commit/cce72e8860e207b5be84acb82c75f0e644a60e29#diff-0b8352c8cc3b9454ebcd4300ca2aaf43f3b09d929fd91b776ac2c9b55f781b53R3-R235), and from [FreeBSD 8](https://github.com/rust-lang/libc/pull/937/files#diff-a828924ffbc30c564e49bc3652ffa5375a123671b228fc99b5775a29cff6eea5R449) for [fstatat(2)](https://github.com/freebsd/freebsd-src/commit/ba2983e5b3676ed66e051aa06281c8c069f90cb0#diff-8758e759f667eb55116ff36d588e0b0d2e2eafc9f72a53d6f92464c83b3b61eaR342). There may be some from 10 or 11 somewhere too. If we _only_ drop all of these then Rust will no longer work on FreeBSD 12+.

> Isn't it finally time to drop FreeBSD 10 support and update the standard library's ABI to FreeBSD 12

- [ ] This only makes sense if the official Rust, and crate, builders are 12, and `rustup` and git builds on FreeBSD continue to work without external hacks.

It appears the builders are still targeting 11.
- https://github.com/rust-lang/rust/blob/0132f8258ae0fbc4f2b461b28d510222d22aa979/src/ci/docker/host-x86_64/dist-x86_64-freebsd/Dockerfile#L29-L32
- https://github.com/rust-lang/rust/blob/0132f8258ae0fbc4f2b461b28d510222d22aa979/src/ci/docker/scripts/freebsd-toolchain.sh#L8

> But FreeBSD 10 went EoL on October 31, 2018. And now FreeBSD 11 will itself be EoL on September 30, 2021. 

"10 (or 11) being EOL" is not very relevant here as noted about the use of 7.0 and 8.0 symbol versions. The reasoning for this (https://github.com/rust-lang/libc/pull/937, https://github.com/rust-lang/libc/pull/936 https://github.com/rust-lang/libc/pull/721) was never about 10 or 11. From what I recall it was because the Rust builder (on FreeBSD 10) was generating unversioned symbol usage that [broke on **12**](https://github.com/alacritty/alacritty/issues/618) with ABI changes there. Rust only knows about the target OS and arch, not release. So the "10/11" compat was _for 12_. FreeBSD ports had [hacks](https://github.com/johalun/rust-ino64) to generate a "working" `rust` binary but as soon as a crate was used that touched any of the structs from that PR it would blow up.  [[1]](https://github.com/rust-lang/rust/issues/42681) [[2]](https://github.com/rust-lang/libc/issues/570#issuecomment-493184774) [[3]](https://github.com/jnr/jnr-posix/issues/126) [[4]](https://github.com/freebsd/freebsd-ports/commit/87fdd25c8ee1cca5eb685200610bb19ef0b43d82). Using `rustup` or building rust from git, or using the bootstrap, also would not work as they lacked the patches we were carrying.

> Rust's standard library currently uses a FreeBSD 10 compatible ABI

I agree the symbols targetted currently are for "10" (some are 7+) (from [libc basically bundling FreeBSD 11 headers](https://github.com/rust-lang/libc/issues/570#issuecomment-298681916)), but as far as I know the majority of symbols are used unversioned  which is why the problem was present in the first place. That is, I believe that at that time if we purposely bound every symbol to the version of the symbol for that prototype that was copied into the libc header then on 12 it would not have been crashing (unless using an external C library). (Maybe that's a simple compromise regarding #570 but I haven't thought deeply on it.)

Rust should not have any blanket policy of dropping compat/versioned symbols for "EOL" releases - they may be there for a reason that has nothing to do with the specific version. FreeBSD has symbols from 7 and earlier that are still used today and may be bound to at some point. I just want to be sure this is clear as I suspect some of the votes here from fellow FreeBSD devs may be policy votes, rather than informed technical votes. Changing what version is *targeted* is fine regarding EOL, but symbols are a more nuanced discussion.

Having said all of that, allowing interoperability between rust and external libraries using FreeBSD's current ABI is important and the reasoning for this compat is not there anymore _assuming the official Rust builders are upgraded to 12_. But someone needs to consider every piece involved in the ecosystem here before rushing a change in because of a date change. 11 being EOL has little to no effect on anything for Rust beyond interoperability.

We really need https://github.com/rust-lang/libc/issues/570 dealt with to have a target containing the FreeBSD release (with proper header sysroots for cross builds or continued FFI use) or libc flags or special crates or whatever. Either way the real problem is the [need for investment to solve and maintain the solution](https://github.com/rust-lang/rfcs/pull/2048#issuecomment-718122987), and how that would need to be mimiced for other platforms to be consistent.

**Comment by asomers** [comment]
@bdrewery, there is already a PR to update the builders: #89083 .  That should take care of rustup.  And regarding the symbol versions, let me summarize for the benefit of everybody new to this saga:

As we discussed elsewhere, FreeBSD uses ELF symbol versioning, allowing it to occasionally modify libc functions like readdir and stat in a way that is backwards compatible at both the source and binary level.  But Rust uses FFI.  That's nice for cross-compiling, but it forces Rust to lock its libc to one particular FreeBSD version.  For whatever reason, Rust chose 11.  The problem with 12+ that bdrewery mentioned is that prior to FreeBSD 12, Rust was linking to symbols such as `stat` with the default link name "stat".  But FreeBSD 12 changed the signature of stat and renamed the old function "stat@FBSD_1.0".  That meant that Rust's libc was briefly trying to use FreeBSD 11's stat signature with FreeBSD 12's stat function.  Fixing the problem was a matter of specifying the correct link name, as shown in the PRs linked above.

So dropping FreeBSD 10 and 11 support will require removing the FreeBSD 11 link names as well as updating the structure definitions.  Most of the new structure definitions are already in libc.  The rest can be added quickly.

**Comment by unrelentingtech** [comment]
> FreeBSD 12 changed the signature of stat and renamed the old function "stat@FBSD_1.0"

nitpick: to be clear, 12 introduced the `@FBSD_1.5` versions, but the old ones were always (okay, since 7.0) `@FBSD_1.0`. The "renaming" effect is virtual, just a consequence of unversioned symbols pointing to the latest version.

> Rust should not have any blanket policy of dropping compat/versioned symbols for "EOL" releases

I don't think anyone has ever suggested going back to unversioned symbols?

Perhaps this issue should've been called "Update FreeBSD support to use 64-bit inodes, available since FreeBSD 12".

**Comment by bdrewery** [comment]
> > Rust should not have any blanket policy of dropping compat/versioned symbols for "EOL" releases
> 
> I don't think anyone has ever suggested going back to unversioned symbols?
> 

I was more responding to other votes with a strong bias of knowing how EOL policy is handled in FreeBSD. That is, I think FreeBSD is too harsh and kneejerk when EOL comes up. Things are deleted because of an arbitrary date without consideration of what the risk actually is. This is probably the reason for my long response as it seemed like a bandwagon EOL party when I believe it was missed that the problem was never 10 or 11 but 12.

Going to unversioned is exactly what is proposed here isn't it? I am not against that anyhow.

**Comment by emaste** [comment]
Indeed - the old version in 11.x is `stat@FBSD_1.0`, and anything linked for FreeBSD 11.x should reference that. You can examine the symbol table of an 11.x binary to see this, e.g.

```
ref11-amd64% readelf -s /usr/bin//stat | grep -w stat
    13: 0000000000000000    17 FUNC    GLOBAL DEFAULT  UND stat@FBSD_1.0 (2)
```

> the majority of symbols are used unversioned

To be clear, is this now fixed and all references are to the versioned symbols?

**Comment by asomers** [comment]
> To be clear, is this now fixed and all references are to the versioned symbols?
All references in Rust's libc library to symbols that have multiple versions in FreeBSD 11 or later are versioned.  References to symbols with a single version, or symbols that haven't had a new version since 11, are unversioned.

**Comment by unrelentingtech** [comment]
One bonus reason to hurry the transition: the default configuration for riscv64 currently does not enable `COMPAT_*` flags, so no pre-ino64 ABI there. [The riscv64 port is running into this](https://reviews.freebsd.org/D32881).

**Comment by bdrewery** [comment]
What's the holdup here? Does someone just need to submit a PR? I pushed in a lot of this compat to get things working but am not a fan of it.

**Comment by nikic** [comment]
FYI https://github.com/rust-lang/rust/pull/97944 landed, which switches CI to build rustc for FreeBSD 12 rather than 11. (Not entirely sure how that relates to this issue, but it seemed relevant to mention.)

**Comment by asomers** [comment]
Now that rustc is using FreeBSD 12 in CI, the next step is top drop FreeBSD 10 support from libc.  We'll have to rebuild rustc to test that change, but no other crates.  After that's done we can separately drop FreeBSD 11 support.  That might require a little more testing.

**Comment by m-ou-se** [comment]
Let's do a libs+compiler FCP for dropping both FreeBSD 10 and (perhaps a little later, after testing) FreeBSD 11:

@rfcbot merge

**Comment by nagisa** [comment]
FWIW given the @cemeyer’s comment, I think they were suggesting that support for 11 is dropped alongside this version going EOL upstream – which it already has. In fact, [so has 12 at this point](https://www.freebsd.org/releases/).

**Comment by asomers** [comment]
> FWIW given the @cemeyer’s comment, I think they were suggesting that support for 11 is dropped alongside this version going EOL upstream – which it already has. In fact, [so has 12 at this point](https://www.freebsd.org/releases/).

Err, early point releases of 12 are EoL, but the 12 series as a whole is not.  12.3 is still supported, and 12.4 will be released soon.  But yes, we are suggesting dropping support for both 10 and 11.

**Comment by emaste** [comment]
For reference, 12.x is supported to end of 2023.
https://www.freebsd.org/security/#sup

**Comment by laffer1** [comment]
Note that FreeBSD isn't the only consumer of the FreeBSD 11 "support" in rust.  MidnightBSD piggybacks off this since we've had trouble with upstreaming to LLVM.   The next major MidnightBSD release (3.0) should be compatible with FreeBSD 12 support, though.  I'd ask that you guys not drop FreeBSD 11 support until the beginning of next year, at least.  We should have a release out by then.

**Comment by cuviper** [comment]
RE waiting until next year -- If we merged this change *today*, it wouldn't reach stable until January 26.

https://forge.rust-lang.org/#current-release-versions

**Comment by mjguzik** [comment]
Hello, another FreeBSD developer here.

Anything I can help with to get this sorted out? @asomers?

**Comment by asomers** [comment]
@mjguzik  If I understand correctly, we need approval from one of either @davidtwco @nikomatsakis or @pnkfelix .  Then we can:
* Submit a PR to rust-lang/libc that drops support for FreeBSD 10.  Testing that will require rebuilding the rust toolchain with the patched libc, which I don't know how to do.
* Submit a subsequent PR to rust-lang/libc that drops support for FreeBSD 11 and changes the ABI to FreeBSD 12.  This may break the build for other crates, for example nix-rust/nix.  I don't know the best way to deal with that.

**Comment by mjguzik** [comment]
ok, then bonus question: I want to add copy_file_range, which is only available since 13, so 12 would have to be excluded in some manner. From a grep I found one can use #[cfg(freebsd12)] . Should the older releases not be officially dropped, can this roll with a not freebsd10/freebsd11/freebsd12 check instead (don't know how to write in in rust)?

**Comment by asomers** [comment]
> ok, then bonus question: I want to add copy_file_range, which is only available since 13, so 12 would have to be excluded in some manner. From a grep I found one can use #[cfg(freebsd12)] . Should the older releases not be officially dropped, can this roll with a not freebsd10/freebsd11/freebsd12 check instead (don't know how to write in in rust)?

Ahh, easy question.  The answer is, `copy_file_range` is already available.  See https://github.com/rust-lang/libc/pull/3023 .

**Comment by asomers** [comment]
> Ahh, easy question. The answer is, `copy_file_range` is already available. See [rust-lang/libc#3023](https://github.com/rust-lang/libc/pull/3023) .

Err, unless you meant to employ `copy_file_range` in `std::fs::copy`.  I don't think we can do that in the standard library until Rust drops FreeBSD 12 support.  Maybe, if you want to somehow use `dlsym` to look for the function at runtime.

**Comment by mjguzik** [comment]
That is what I meant. I thought the cfg thing is an ifdef-equivalent, meaning it would correctly compile the usage when built on 13 and later and *omit* it otherwise, which maintains all the compat you normally ask for.

**Comment by asomers** [comment]
> That is what I meant. I thought the cfg thing is an ifdef-equivalent, meaning it would correctly compile the usage when built on 13 and later and _omit_ it otherwise, which maintains all the compat you normally ask for.

It's not quite an ifdef equivalent.  Outside of CI, there's nothing that will ever set the `freebsd13` flag in a libc build.  Due to Rust's heavy emphasis on cross-compiling, nobody's yet come up with a good way to do that in such a widely used crate.  See discussion at https://github.com/rust-lang/libc/issues/570 .

**Comment by mjguzik** [comment]
welp, sigh

Linux support tries to work around stuff by handling ENOSYS at runtime, but we can't do that on FreeBSD due to SIGSYS.

While I'm not fond of this, how about pessimizing binary startup with fetching kern.osreldate. Then io::copy (and possibly many others) can branch on it. Does not solve binary compat, but it is always *something*. I guess for an initial version I can pessimize it just for code using the copy routine.

**Comment by asomers** [comment]
Actually, the situation is both better and worse than that.  Linux can optimistically attempt copy_file_range and check for ENOSYS because it makes a direct syscall.  With FreeBSD, we need to link to libc.so, so we'll get a link-time failure if you invoke the program on a system without copy_file_range.  OTOH, if the link succeeds, then there's virtually no chance that the kernel support won't be present, so we don't need to check for ENOSYS.  That's why I think that a dlsym-based approach would work.  But since it's unrelated to dropping 10 and 11 support, maybe you should open a separate issue?

**Comment by bsdimp** [comment]
> how about pessimizing binary startup with fetching kern.osreldate.

getosreleasedate() will grab this information from the ELF AUX vector w/o a trip into the kernel for any kernel revision that's relevant for this discussion.

**Comment by mjguzik** [comment]
> > how about pessimizing binary startup with fetching kern.osreldate.
> 
> getosreleasedate() will grab this information from the ELF AUX vector w/o a trip into the kernel for any kernel revision that's relevant for this discussion.

that's what i wrote: 
> there is osreldate in the auxv vector, so no need to make any syscalls.

anyhow i'll be posting a new PR few days from now

**Comment by emaste** [comment]
Should this issue be closed now?

**Comment by RalfJung** [comment]
>  Outside of CI, there's nothing that will ever set the freebsd13 flag in a libc build.

That's not true any more, this is now set by the build script. That doesn't work well for cross-builds but at least when building libc on freebsd, it picks the freebsd API matching the current OS.

> Should this issue be closed now?

libc still has support for freebad10 and 11 (and even defaults to 11 when the version can't be detected). But maybe that should be a separate issue? Not sure.

**Comment by asomers** [comment]
> That's not true any more, this is now set by the build script. That doesn't work well for cross-builds but at least when building libc on freebsd, it picks the freebsd API matching the current OS.

Where does it do that? In the main branch,  libc's build script still sets freebsd11 unless libc_ci is set.

**Comment by RalfJung** [comment]
Ah, I missed the `if libc_ci` match guards. Though strangely under `rustc_dep_of_std` it will use freebsd12 if `which_freebsd` returns `Some` and freebsd11 otherwise...

**Comment by tgross35** [comment]
Could somebody familiar with this target (@asomers perhaps) create a platform doc page listing the supported version? Just the template here https://github.com/rust-lang/rust/tree/0f26ee4fd95a1c046582dfb18892f520788e2c2c/src/doc/rustc/src/platform-support and linking the targets to it here https://github.com/rust-lang/rust/blob/0f26ee4fd95a1c046582dfb18892f520788e2c2c/src/doc/rustc/src/platform-support.md. It currently isn't very easy to find supported versions, this is usually a good place for it.

**Comment by clausecker** [comment]
@tgross35 The oldest FreeBSD release we currently support is 13.3.  See [this page](https://www.freebsd.org/releases/) for details.  Is this the information you are looking for?

**Comment by tgross35** [comment]
Do you mean 13.3 is the oldest FreeBSD release that FreeBSD supports, or that is the oldest FreeBSD that `rustc` is expected to build on? I assumed `rustc` is somewhere around 12 based on https://github.com/rust-lang/rust/pull/114521. I was just asking for that (and any other relevant information for building on FreeBSD) to be put into some platform docs so it's a bit easier to find.

**Comment by clausecker** [comment]
@tgross35  The former.  As for the latter, I do not have information on that.

**Comment by emaste** [comment]
> Could somebody familiar with this target create a platform doc page listing the supported version

Supported FreeBSD releases are listed at https://www.freebsd.org/security/#sup

Looking at the existing we have:
x86_64-unknown-freebsd - Rust tier 2 w/ host tools
i686-unknown-freebsd - Rust tier 2 w/o host tools
aarch64-unknown-freebsd - Rust tier 3 (same for all following)
armv6-unknown-freebsd
armv7-unknown-freebsd
powerpc64-unknown-freebsd
powerpc64le-unknown-freebsd
powerpc-unknown-freebsd
riscv64gc-unknown-freebsd

We also have support for powerpcspe, and 32- and 64-bit big- and little-endian MIPS. MIPS is only supported through 13.x.

x86_64 (aka amd64) and aarch64 are the two Tier-1 architectures for us.

I agree that based on #114521 the minimum FreeBSD version supported by rustc should be 12.x. There's nothing standing in the way of increasing this to 13.x, if someone has a reason to do so.

I'm not yet sufficiently familiar with Rust to populate the doc pages; I'll see if I can work with @asomers to do so (or find someone else).

**Comment by RalfJung** [comment]
As far as I can tell from [this code](https://github.com/rust-lang/libc/blob/e13cde2dff7d4965bfb12577f35d488ec3ba8a22/build.rs#L61), we already require FreeBSD 12 when building std. (It's somewhat concerning that this is configured in the libc build script though, shouldn't rustc bootstrap be in charge of setting the right version?)

So -- is there anything left to do for this issue? Cc @asomers @MikaelUrankar

**Comment by emaste** [comment]
There's still some code in libc/build.rs for FreeBSD 10 and 11 that could be updated or removed, e.g.
 
```
    match which_freebsd {
        x if x < 10 => panic!("FreeBSD older than 10 is not supported"),
        10 => set_cfg("freebsd10"),
        11 => set_cfg("freebsd11"),
```

But I think the main point of this issue (using at least FreeBSD 12's ABI, in particular 64-bit inodes) is complete.

**Comment by asomers** [comment]
There's still one outstanding problem, @emaste @RalfJung : #132185 .  When we switched CI to build Rust on FreeBSD 13.4, it produced binaries that segfault when they try to print a stack trace.  We've never tracked down the root cause of that bug.  It's probably an LLVM bug.

**Comment by emaste** [comment]
@asomers #132185 is now closed; is the issue solved?

**Comment by asomers** [comment]
No, actually it isn't @emaste .  I still see occasional segfaults in CI for FreeBSD 13, especially 13.4 and 13.5.  But never on 14.  I haven't done any more work to investigate this issue for months.

**Comment by RodBurman** [comment]
It seems a bit odd discussing dropping support for BSD 13 in an issue called "Drop support for FreeBSD 10 and 11 from std", should we close this issue? Or at least rename it to something like "keeping STD aligned with supported versions of BSD".

**Comment by emaste** [comment]
> It seems a bit odd discussing dropping support for BSD 13 in an issue called "Drop support for FreeBSD 10 and 11 from std"

True. There's also somewhat less motivation to drop support for FreeBSD 12 or later. We wanted to drop support for versions earlier than 12, as they lacked support for 64-bit inodes.
